examples/retrieval/retrieval.cpp

```diff
@@ -0,0 +1,358 @@
+                retrieval_params_print_usage(argc, argv, gpt_params, retrieval_params);
+                exit(1);
+            }
+            // option was parsed by gpt_params_find_arg
+        } else if (arg == "--context-files") {
+            if (++i >= argc) {
+                fprintf(stderr, "error: missing argument for --context-files\n");
+                retrieval_params_print_usage(argc, argv, gpt_params, retrieval_params);
+                exit(1);
+            }
+            while(true) {
+                std::ifstream file(argv[i]);
+                if (!file) {
+                    fprintf(stderr, "error: failed to open file '%s'\n", argv[i]);
+                    retrieval_params_print_usage(argc, argv, gpt_params, retrieval_params);
+                    exit(1);
+                }
+                // store the external file name in params
+                retrieval_params.context_files.push_back(argv[i]);
+                if (i + 1 >= argc || argv[i + 1][0] == '-') {
+                    break;
+                }
+                i++;
+            }
+        } else if (arg == "--chunk-size") {
+            if (++i >= argc) {
+                fprintf(stderr, "error: missing argument for --chunk-size\n");
+                retrieval_params_print_usage(argc, argv, gpt_params, retrieval_params);
+                exit(1);
+            }
+            retrieval_params.chunk_size = std::stoi(argv[i]);
+        } else if (arg == "--chunk-separator") {
+            if (++i >= argc) {
+                fprintf(stderr, "error: missing argument for --chunk-separator\n");
+                retrieval_params_print_usage(argc, argv, gpt_params, retrieval_params);
+                exit(1);
+            }
+            retrieval_params.chunk_separator = argv[i];
+        } else {
+            // unknown argument
+            fprintf(stderr, "error: unknown argument: %s\n", arg.c_str());
+            retrieval_params_print_usage(argc, argv, gpt_params, retrieval_params);
+            exit(1);
+        }
+        i++;
+    }
+}
+
+struct chunk {
+    // filename
+    std::string filename;
+    // original file position
+    size_t filepos;
+    // original text data
+    std::string textdata = "";
+    // tokenized text data
+    std::vector<llama_token> tokens;
+    // embedding
+    std::vector<float> embedding;
+};
+
+// chunk file data to chunks of size >= chunk_size
+// chunk_separator is the separator between chunks
+static std::vector<chunk> chunk_file(const std::string filename, int chunk_size, std::string chunk_separator) {
+    std::vector<chunk> chunks;
+    std::ifstream f(filename.c_str());
+
+    if (!f.is_open()) {
+        fprintf(stderr, "Error: could not open file %s\n", filename.c_str());
+        return chunks;
+    }
+
+    chunk current_chunk;
+    char buffer[100];
+    int64_t filepos = 0;
+    std::string current = "";
+    while (f.read(buffer, 100)) {
```

Comment:
This approach may not work if the `chunk_separator` has 2 bytes (or more). That's because half of `chunk_separator` maybe in on buffer, while the other half maybe in the next buffer.

It's safer and more simple to just read in the whole file: https://www.tutorialspoint.com/Read-whole-ASCII-file-into-Cplusplus-std-string

Reply:
The problem that you stated probably won't happen because we accumulate the readed content to `current` variable and run `find` over it

Reply:
I see, OK to me. This approach is not very efficient in term of complexity (because `current.find` is called too often), but we can improve it in the future if needed.

Here can we make buffer larger (and become a power of 2), for example `char buffer[1024];` ?